runtime/wasm/src/host_exports.rs

```diff
@@ -335,6 +343,48 @@ where
         )
     }
 
+    // Read the IPFS file `link`, split it into JSON objects, and invoke
+    // the exported function `callback` on each JSON object. The successful
+    // return value contains all entity operations that were produced by the
+    // callback invocations. Each invocation of `callback` happens in its own
+    // instance of a WASM module, which is identical to `module` when it was
+    // first started.
+    pub(crate) fn ipfs_map(
+        &self,
+        module: &WasmiModule<E, L, S, U>,
+        link: String,
+        callback: &str,
+        flags: Vec<String>,
+    ) -> Result<Vec<EntityOperation>, HostExportError<impl ExportError>> {
+        const JSON_FLAG: &str = "json";
+        if !flags.contains(&JSON_FLAG.to_string()) {
+            return Err(HostExportError(format!("Flags must contain 'json'")));
+        }
+
+        let valid_module = module.valid_module.clone();
+        let ctx = module.ctx.clone();
+        let callback = callback.to_owned();
+
+        let results = self.block_on(
+            self.link_resolver
+                .json_stream(&Link { link })
+                .and_then(|stream| {
+                    stream
+                        .and_then(move |v| {
+                            let module = WasmiModule::from_valid_module_with_ctx(
+                                valid_module.clone(),
+                                ctx.clone(),
+                            )?;
+                            module.handle_json_callback(&*callback, &v)
+                        })
+                        .collect()
+                })
+                .map_err(|e| HostExportError(e.to_string())),
+        )?;
+        // Collect all results into one Vec
+        Ok(results.into_iter().flatten().collect())
```

Comment:
I am confused by your 'fail fast' comment: I thought the `collect` on line 380 would abort processing the stream as soon as an error happens, and we would abort the whole `block_on` as soon as the `and_then` on line 373 encounters an error. Is that not correct?

Reply:
You're right of course, I misunderstood. I think what mislead me is that because of the name I thought `results` held was a vector of `Result`, so maybe renaming that variable would be good.

Reply:
I renamed it to `operations`